build/xtask/src/sizes.rs

```diff
@@ -79,233 +68,235 @@ pub fn run(
+        o => bail!("Invalid Object {:?}", o),
     };
 
-    let check_task = move |name: &str,
-                           stacksize: u32,
-                           requires: &IndexMap<String, u32>|
-          -> anyhow::Result<(
-        IndexMap<String, u64>,
-        Vec<Vec<(String, u32, u64)>>,
-    )> {
-        let mut suggestions = Vec::new();
-        let mut sizes = IndexMap::new();
-
-        let mut elf_name = dist_dir.clone();
-        elf_name.push(name);
-        let buffer = std::fs::read(elf_name)?;
-        let elf = match Object::parse(&buffer)? {
-            Object::Elf(elf) => elf,
-            o => bail!("Invalid Object {:?}", o),
-        };
-
-        let mut memory_sizes = IndexMap::new();
-        for phdr in &elf.program_headers {
-            if let Some(vregion) = output_region(phdr.p_vaddr) {
-                *memory_sizes.entry(vregion).or_default() += phdr.p_memsz;
-            }
-            // If the VirtAddr disagrees with the PhysAddr, then this is a
-            // section which is relocated into RAM, so we also accumulate
-            // its FileSiz in the physical address (which is presumably
-            // flash).
-            if phdr.p_vaddr != phdr.p_paddr {
-                let region = output_region(phdr.p_paddr).unwrap();
-                *memory_sizes.entry(region).or_default() += phdr.p_filesz;
-            }
+    // We can't naively add up section sizes, since there may be gaps left
+    // by alignment requirements.  Instead, we track the min and max bounds
+    // within each memory region (flash, RAM, etc), then extract the sizes
+    // afterwards.
+    let mut memory_sizes = IndexMap::new();
+    let mut record_size = |start, size| {
+        if let Some(region) = toml.output_region(start) {
+            let end = start + size;
+            let r = memory_sizes.entry(region).or_insert_with(|| start..end);
+            r.start = r.start.min(start);
+            r.end = r.end.max(end);
+            true
+        } else {
+            false
         }
-        *memory_sizes.entry("ram").or_default() += stacksize as u64;
-
-        let mut my_suggestions = Vec::new();
-
-        for (mem_name, used) in memory_sizes {
-            if let Some(&size) = requires.get(mem_name) {
-                sizes.insert(mem_name.to_string(), used);
-                let suggestion = suggest(used);
-                if suggestion < size as u64 {
-                    my_suggestions.push((
-                        mem_name.to_string(),
-                        size,
-                        suggestion,
-                    ));
-                }
-            } else {
-                assert!(used == 0);
+    };
+    for phdr in &elf.program_headers {
+        record_size(phdr.p_vaddr, phdr.p_memsz);
+
+        // If the VirtAddr disagrees with the PhysAddr, then this is a
+        // section which is relocated into RAM, so we also accumulate
+        // its FileSiz in the physical address (which is presumably
```

Comment:
is `FileSiz` a typo? Shouldn't it be "file size" or `FileSz` or something? an extremely minor thing.

Reply:
heh, the names in this section based on the output of `readelf`, so it's intentional
```
Program Headers:
  Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
  LOAD           0x000120 0x08039400 0x08039400 0x00060 0x00060 R E 0x20
  LOAD           0x000180 0x08039460 0x08039460 0x00020 0x00020 R E 0x20
  GNU_STACK      0x000000 0x00000000 0x00000000 0x00000 0x00000 RW  0x0
```